iroh-router/src/router.rs

```diff
@@ -0,0 +1,211 @@
+    // - `Shared` so we can `task.await` from all `Node` clones
+    // - `MapErr` to map the `JoinError` to a `String`, because `JoinError` is `!Clone`
+    // - `AbortOnDropHandle` to make sure that the `task` is cancelled when all `Node`s are dropped
+    //   (`Shared` acts like an `Arc` around its inner future).
+    task: Shared<MapErr<AbortOnDropHandle<()>, JoinErrToStr>>,
+    cancel_token: CancellationToken,
+}
+
+type JoinErrToStr = Box<dyn Fn(JoinError) -> String + Send + Sync + 'static>;
+
+impl Router {
+    pub fn builder(endpoint: Endpoint) -> RouterBuilder {
+        RouterBuilder::new(endpoint)
+    }
+
+    /// Returns a protocol handler for an ALPN.
+    ///
+    /// This downcasts to the concrete type and returns `None` if the handler registered for `alpn`
+    /// does not match the passed type.
+    pub fn get_protocol<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        self.protocols.get_typed(alpn)
+    }
+
+    pub fn endpoint(&self) -> &Endpoint {
+        &self.endpoint
+    }
+
+    pub async fn shutdown(self) -> Result<()> {
+        // Trigger shutdown of the main run task by activating the cancel token.
+        self.cancel_token.cancel();
+
+        // Wait for the main task to terminate.
+        self.task.await.map_err(|err| anyhow!(err))?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct RouterBuilder {
+    endpoint: Endpoint,
+    protocols: ProtocolMap,
+}
+
+impl RouterBuilder {
+    pub fn new(endpoint: Endpoint) -> Self {
+        Self {
+            endpoint,
+            protocols: ProtocolMap::default(),
+        }
+    }
+
+    pub fn accept(mut self, alpn: Vec<u8>, handler: Arc<dyn ProtocolHandler>) -> Self {
+        self.protocols.insert(alpn, handler);
+        self
+    }
+
+    /// Returns the [`Endpoint`] of the node.
+    pub fn endpoint(&self) -> &Endpoint {
+        &self.endpoint
+    }
+
+    /// Returns a protocol handler for an ALPN.
+    ///
+    /// This downcasts to the concrete type and returns `None` if the handler registered for `alpn`
+    /// does not match the passed type.
+    pub fn get_protocol<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
+        self.protocols.get_typed(alpn)
+    }
+
+    pub async fn spawn(self) -> Result<Router> {
+        // Update the endpoint with our alpns.
+        let alpns = self
+            .protocols
+            .alpns()
+            .map(|alpn| alpn.to_vec())
+            .collect::<Vec<_>>();
+
+        let protocols = Arc::new(self.protocols);
+        if let Err(err) = self.endpoint.set_alpns(alpns) {
+            shutdown(&self.endpoint, protocols.clone()).await;
+            return Err(err);
+        }
+
+        let mut join_set = JoinSet::new();
+        let endpoint = self.endpoint.clone();
+        let protos = protocols.clone();
+        let cancel = CancellationToken::new();
+        let cancel_token = cancel.clone();
+
+        let fut = async move {
+            let protocols = protos;
```

Comment:
```rust
let fut = {
    let protocols = protocol.clone();
    async move { ... }
};
```
Avoids this awkward dance of `let protos = protocols.clone(); let protocols = protos;` and thus is a style I prefer.  (you can chuck all the other clones there as well)

Reply:
but it adds another level of indentation to the whole block, which I think is worse